### DISCUSSION TITLE
Prowler JSON report structure?

### DISCUSSION THREAD

**Discussion by JWebDev** [prowler-cloud/prowler#1308] [Q&A] [answered]
Hi. 
I am currently unable to generate a report and look at its structure.
Also, I can’t find what the json report looks like in the documentation.

Could you tell me where I can see structure of json report?

Thank you!

**Comment by toniblyx** [comment, marked as answer]
Hi @JWebDev can you give us more details on how are you running prowler and what version are you using?
The easiest way to get json output reports is with `./prowler -M json` for all checks in all regions, if you want to narrow down for tests do something like `./prowler -M json,html,csv -f us-east-1 -c check11,check12`

**Reply by JWebDev** [reply]
Hi, @toniblyx Thanks for the quick response.
That is why I am asking the question. At the moment I can't start prowler and generate report. But I want to know report structure.
If you don't have documentation on what the report looks like, could you post some test case here.
Thanks. ))

**Reply by toniblyx** [reply]
ok, I didn't get what you mean. 
We have the following for our json output (sample): `./prowler -M json`
```
{
  "Profile": "ENV",
  "Account Number": "1111111111111",
  "Control": "[check14] Ensure access keys are rotated every 90 days or less",
  "Message": "us-west-2: terraform-user has not rotated access key 2 in over 90 days",
  "Severity": "Medium",
  "Status": "FAIL",
  "Scored": "",
  "Level": "CIS Level 1",
  "Control ID": "1.4",
  "Region": "us-west-2",
  "Timestamp": "2022-05-18T10:33:48Z",
  "Compliance": "ens-op.acc.1.aws.iam.4 ens-op.acc.5.aws.iam.3",
  "Service": "iam",
  "CAF Epic": "IAM",
  "Risk": "Access keys consist of an access key ID and secret access key which are used to sign programmatic requests that you make to AWS. AWS users need their own access keys to make programmatic calls to AWS from the AWS Command Line Interface (AWS CLI)- Tools for Windows PowerShell- the AWS SDKs- or direct HTTP calls using the APIs for individual AWS services. It is recommended that all access keys be regularly rotated.",
  "Remediation": "Use the credential report to  ensure  access_key_X_last_rotated  is less than 90 days ago.",
  "Doc link": "https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_getting-report.html",
  "Resource ID": "terraform-user",
  "Account Email": "",
  "Account Name": "",
  "Account ARN": "",
  "Account Organization": "",
  "Account tags": ""
}
```
we also support ASFF for AWS Security Hub which is this one: `./prowler -M json-asff`
```
{
  "SchemaVersion": "2018-10-08",
  "Id": "prowler-1.4-1111111111111-us-west-2-us-west-2_terraform-user_has_not_rotated_access_key_2_in_over_90_days",
  "ProductArn": "arn:aws:securityhub:us-west-2::product/prowler/prowler",
  "RecordState": "ACTIVE",
  "ProductFields": {
    "ProviderName": "Prowler",
    "ProviderVersion": "2.9.0-13April2022",
    "ProwlerResourceName": "terraform-user"
  },
  "GeneratorId": "prowler-check14",
  "AwsAccountId": "1111111111111",
  "Types": [
    "ens-op.acc.1.aws.iam.4 ens-op.acc.5.aws.iam.3"
  ],
  "FirstObservedAt": "2022-05-18T10:33:48Z",
  "UpdatedAt": "2022-05-18T10:33:48Z",
  "CreatedAt": "2022-05-18T10:33:48Z",
  "Severity": {
    "Label": "MEDIUM"
  },
  "Title": "iam.[check14] Ensure access keys are rotated every 90 days or less",
  "Description": "us-west-2: terraform-user has not rotated access key 2 in over 90 days",
  "Resources": [
    {
      "Type": "AwsIamUser",
      "Id": "terraform-user",
      "Partition": "aws",
      "Region": "us-west-2"
    }
  ],
  "Compliance": {
    "Status": "FAILED",
    "RelatedRequirements": [
      "ens-op.acc.1.aws.iam.4 ens-op.acc.5.aws.iam.3"
    ]
  }
}
```

in the coming version 3 our json and cvs formats will come with more keys and comprehensive content.

**Reply by JWebDev** [reply]
@toniblyx Thanks!
That's what I need.
As far as I understand, this is an object for one check. What if there are multiple objects? Will there be an array of objects?

```
[{
 "Profile": "ENV",
  "Account Number": "1111111111111",
  .......
},{
 "Profile": "ENV",
  "Account Number": "1111111111111",
  .......
}]
```

Could you please add a description of the report structure to the documentation? Current or new (version 3), with new keys?

Thanks a lot!

**Reply by toniblyx** [reply]
not a check but a finding, that is for just one finding, usually you will get multiple objects, one per finding.

**Reply by JWebDev** [reply]
Sorry I expressed it incorrectly.
If you look at the documentation `https://github.com/prowler-cloud/prowler#usage`
`./prowler -M json`

There must be several findings, so the structure is correct? Will there be an array of objects?
So as I described above?

I ask again because I want to understand correctly. Thanks for understanding.

**Reply by toniblyx** [reply]
they are separate objects in the same json file.

**Reply by toniblyx** [reply]
in v3 will be a list of objects but now in v2 are separated objects.

**Reply by JWebDev** [reply]
Understood. And the last mini question.
When to expect the release of version 3?

Please add the report structure/description to the documentation. Thanks!

**Reply by toniblyx** [reply]
as soon as we finish porting all checks it will be released officially, probably by the end of September, not written in stone though.

We will, thanks for your feedback, much appreciated!

**Reply by JWebDev** [reply]
Ahhh, so very soon. ))))
Thanks for the info.

**Comment by toniblyx** [comment]
Documentation updated with these details: https://github.com/prowler-cloud/prowler#output-formats
